### PR TITLE
Respect `pyproject.toml` credentials from user-provided requirements

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

When syncing a lockfile, we need to respect credentials defined in the `pyproject.toml`, even if they won't be used for resolution. Unfortunately, this includes credentials in `tool.uv.sources`, `tool.uv.dev-dependencies`, `project.dependencies`, and `project.optional-dependencies`.

Closes https://github.com/astral-sh/uv/issues/7453.

### CODE REVIEW

File: crates/uv/src/commands/project/sync.rs [hunk 1]
```diff
@@ -399,3 +405,78 @@ fn apply_editable_mode(
         }),
     }
 }
+
+fn store_credentials_from_workspace(workspace: &Workspace) {
```
Comment (zanieb) [inline crates/uv/src/commands/project/sync.rs:409]: Might be nice to have a doc here

**Review by zanieb — APPROVED** [review]

**Comment by charliermarsh** [comment]
Depressed that exactly one Windows test is overflowing the stack.

File: crates/uv/src/commands/project/sync.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv/src/commands/project/sync.rs:409]: Oh shoot forgot. Will PR separately.

File: crates/uv/tests/common/mod.rs [hunk 2]
```diff
@@ -533,7 +533,7 @@ impl TestContext {
         if cfg!(all(windows, debug_assertions)) {
             // TODO(konstin): Reduce stack usage in debug mode enough that the tests pass with the
             // default windows stack of 1MB
-            command.env("UV_STACK_SIZE", (2 * 1024 * 1024).to_string());
+            command.env("UV_STACK_SIZE", (4 * 1024 * 1024).to_string());
```
Comment (charliermarsh) [inline crates/uv/tests/common/mod.rs:536]: \cc @konstin -- I had to bump this because one test was failing...

**Pull request merged**